bundles/org.openhab.binding.solarman/src/main/java/org/openhab/binding/solarman/internal/typeprovider/SolarmanChannelTypeProvider.java

```diff
@@ -0,0 +1,142 @@
+    private final Map<ChannelTypeUID, ChannelType> channelTypeMap = new ConcurrentHashMap<>();
+
+    @Activate
+    public SolarmanChannelTypeProvider(BundleContext bundleContext) {
+        Collections.list(bundleContext.getBundle().findEntries("/definitions", "*", false)).stream().map(URL::getFile)
+                .map(this::extractInverterDefinitionId).filter(Optional::isPresent).map(Optional::get)
+                .map(this::parseInverterDefinition).forEach(channelTypeMap::putAll);
+    }
+
+    private Map<ChannelTypeUID, ChannelType> parseInverterDefinition(String inverterDefinitionId) {
+        InverterDefinition inverterDefinition = DEFINITION_PARSER.parseDefinition(inverterDefinitionId);
+
+        if (inverterDefinition == null) {
+            logger.error("Unable to parse inverter definition");
+            return Collections.emptyMap();
+        }
+
+        return inverterDefinition.getParameters().stream()
+                .flatMap(parameter -> parameter.getItems().stream().map(item -> {
+                    ChannelTypeUID channelTypeUID = ChannelUtils.computeChannelTypeId(inverterDefinitionId,
+                            parameter.getGroup(), item.getName());
+                    return new AbstractMap.SimpleEntry<>(channelTypeUID, buildChannelType(channelTypeUID, item));
+                })).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private Optional<String> extractInverterDefinitionId(String file) {
+        return Stream.of(file).map(INVERTER_DEFINITION_PATTERN::matcher).filter(Matcher::matches)
+                .map(matcher -> matcher.group(1)).findFirst();
+    }
+
+    public Collection<ChannelType> getChannelTypes(@Nullable Locale locale) {
+        return List.copyOf(this.channelTypeMap.values());
+    }
+
+    @Override
+    public @Nullable ChannelType getChannelType(ChannelTypeUID channelTypeUID, @Nullable Locale locale) {
+        return this.channelTypeMap.get(channelTypeUID);
+    }
+
+    public ChannelType buildChannelType(ChannelTypeUID channelTypeUID, ParameterItem item) {
+        String itemType = getItemType(item);
+
+        StateDescriptionFragmentBuilder stateDescriptionFragmentBuilder = StateDescriptionFragmentBuilder.create()
+                .withPattern(computePatternForItem(item)).withReadOnly(true);
+
+        StateChannelTypeBuilder stateChannelTypeBuilder = ChannelTypeBuilder
+                .state(channelTypeUID, item.getName(), itemType)
+                .withConfigDescriptionURI(URI.create("channel-type-config:solarman:dynamic-channel"))
+                .withDescription(String.format("%s %s", item.getName(), buildRegisterDescription(item)))
+                .withStateDescriptionFragment(stateDescriptionFragmentBuilder.build());
+
+        return stateChannelTypeBuilder.build();
+    }
+
+    private String computePatternForItem(ParameterItem item) {
+        long decimalPoints = 0;
+
+        @Nullable
```

Comment:
item.getScale() is already annotated with `@Nullable` so this should not be needed to annotate here again